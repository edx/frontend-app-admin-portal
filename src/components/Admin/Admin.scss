.row.equal-col-height {
    display: flex;
    flex-wrap: wrap;

    > [class*="col-"] {
        display: flex;
    }

    .card {
        flex: 1;
    }
}
