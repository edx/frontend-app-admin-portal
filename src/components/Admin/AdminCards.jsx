import React from 'react';
import PropTypes from 'prop-types';

import NumberCard from '../../components/NumberCard';

class AdminCards extends React.Component {
  constructor(props) {
    super(props);

    this.cards = {
      numberOfUsers: {
        ref: React.createRef(),
        description: 'total number of learners registered',
        iconClassName: 'fa fa-users',
        actions: [{
          label: 'Which learners are registered but not yet enrolled in any courses?',
          slug: 'registered-unenrolled-learners',
        }],
      },
      enrolledLearners: {
        ref: React.createRef(),
        description: 'learners enrolled in at least one course',
        iconClassName: 'fa fa-check',
        actions: [{
          label: 'How many courses are learners enrolled in?',
          slug: 'enrolled-learners',
        }, {
          label: 'Who is no longer enrolled in a current course?',
          slug: 'enrolled-learners-inactive-courses',
        }],
      },
      activeLearners: {
        ref: React.createRef(),
        description: 'active learners in the past week',
        iconClassName: 'fa fa-eye',
        actions: [{
          label: 'Who are my top active learners?',
          slug: 'learners-active-week',
        }, {
          label: 'Who has not been active for over a week?',
          slug: 'learners-inactive-week',
        }, {
          label: 'Who has not been active for over a month?',
          slug: 'learners-inactive-month',
        }],
      },
      courseCompletions: {
        ref: React.createRef(),
        description: 'course completions',
        iconClassName: 'fa fa-trophy',
        actions: [{
          label: 'How many courses have been completed by learners?',
          slug: 'completed-learners',
        }, {
          label: 'Who completed a course in the past week?',
          slug: 'completed-learners-week',
        }],
      },
    };
  }

  renderCard({ title, cardKey }) {
    const card = this.cards[cardKey];

    return (
      <div
        className="col-xs-12 col-md-6 col-xl-3 mb-3"
        key={cardKey}
      >
        <NumberCard
          id={cardKey}
          title={title}
          description={card.description}
          iconClassName={card.iconClassName}
          detailActions={card.actions}
        />
      </div>
    );
  }

  render() {
    const {
      activeLearners,
      numberOfUsers,
      courseCompletions,
      enrolledLearners,
    } = this.props;

    const data = {
      activeLearners: activeLearners.past_week,
      numberOfUsers,
      courseCompletions,
      enrolledLearners,
    };

    return (
      <div className="row mt-3 equal-col-height">
        {Object.keys(this.cards).map(cardKey =>
          this.renderCard({ title: data[cardKey], cardKey }))}
      </div>
    );
  }
}

AdminCards.propTypes = {
  activeLearners: PropTypes.shape({
    past_week: PropTypes.number.isRequired,
    past_month: PropTypes.number.isRequired,
  }).isRequired,
  numberOfUsers: PropTypes.number.isRequired,
  courseCompletions: PropTypes.number.isRequired,
  enrolledLearners: PropTypes.number.isRequired,
};

export default AdminCards;
