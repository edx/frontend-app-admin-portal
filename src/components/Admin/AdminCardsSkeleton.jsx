import React from 'react';
import Skeleton from 'react-loading-skeleton';

const AdminCardsSkeleton = () => (
  <div
    className="admin-cards-skeleton mb-3 d-md-flex w-100"
  >
    <Skeleton />
    <Skeleton />
    <Skeleton />
    <Skeleton />
  </div>
);

export default AdminCardsSkeleton;
