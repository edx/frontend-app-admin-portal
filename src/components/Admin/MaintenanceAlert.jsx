import React from 'react';
import { Alert } from '@edx/paragon';

import { configuration } from '../../config';

export default () => {
  if (!configuration?.SHOW_LEARNER_REPORT_MAINTENANCE_ALERT) {
    return null;
  }

  return (
    <Alert variant="warning" className="mt-3">
      Learner Progress Report will be down for planned maintenance on Wednesday, December 16th from 10am-12pm EST
    </Alert>
  );
};
