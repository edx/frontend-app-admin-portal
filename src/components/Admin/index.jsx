import React from 'react';
import PropTypes from 'prop-types';
import Helmet from 'react-helmet';
import { MailtoLink, Button } from '@edx/paragon';

import H2 from '../../components/H2';
import Hero from '../../components/Hero';
import NumberCard from '../../components/NumberCard';
import StatusAlert from '../../components/StatusAlert';
import LoadingMessage from '../../components/LoadingMessage';
import CourseEnrollmentsTable from '../../containers/CourseEnrollmentsTable';

import { formatTimestamp } from '../../utils';

import './Admin.scss';

class Admin extends React.Component {
  componentDidMount() {
    const { enterpriseId } = this.props;

    if (enterpriseId) {
      this.props.getDashboardAnalytics(enterpriseId);
    }
  }

  componentDidUpdate(prevProps) {
    const { enterpriseId } = this.props;
    if (enterpriseId && enterpriseId !== prevProps.enterpriseId) {
      this.props.getDashboardAnalytics(enterpriseId);
    }
  }

  hasAnalyticsData() {
    const {
      activeLearners,
      numberOfUsers,
      courseCompletions,
      enrolledLearners,
    } = this.props;
    let hasAnalyticsData = false;

    [activeLearners, courseCompletions, enrolledLearners, numberOfUsers].forEach((item) => {
      hasAnalyticsData = item !== null || hasAnalyticsData;
    });

    return hasAnalyticsData;
  }

  renderErrorMessage() {
    return (
      <StatusAlert
        className={['mt-3']}
        alertType="danger"
        iconClassName={['fa', 'fa-times-circle']}
        title="Unable to load overview"
        message={`Try refreshing your screen (${this.props.error.message})`}
      />
    );
  }

  renderLoadingMessage() {
    return <LoadingMessage className="overview mt-3" />;
  }

  render() {
    const {
      activeLearners,
      numberOfUsers,
      courseCompletions,
      enrolledLearners,
      error,
      loading,
      downloadCsv,
      enterpriseId,
      lastUpdatedDate,
    } = this.props;

    return (
      <div>
        <Helmet>
          <title>Learner and Progress Report</title>
        </Helmet>
        <Hero title="Learner and Progress Report" />
        <div className="container">
          <div className="row mt-4">
            <div className="col">
              <H2>Overview</H2>
              {error && this.renderErrorMessage()}
              {loading && this.renderLoadingMessage()}
              {!loading && !error && this.hasAnalyticsData() &&
                <div className="row mt-3 equal-col-height">
                  <div className="col-xs-12 col-md-6 col-xl-3 mb-3">
                    <NumberCard
                      title={numberOfUsers}
                      description="total number of learners registered"
                      iconClassName="fa fa-users"
                    />
                  </div>
                  <div className="col-xs-12 col-md-6 col-xl-3 mb-3">
                    <NumberCard
                      title={enrolledLearners}
                      description="learners enrolled in at least one course"
                      iconClassName="fa fa-check"
                    />
                  </div>
                  <div className="col-xs-12 col-md-6 col-xl-3 mb-3">
                    <NumberCard
                      title={activeLearners.past_week}
                      description="active learners in the past week"
                      iconClassName="fa fa-eye"
                    />
                  </div>
                  <div className="col-xs-12 col-md-6 col-xl-3 mb-3">
                    <NumberCard
                      title={courseCompletions}
                      description="course completions"
                      iconClassName="fa fa-trophy"
                    />
                  </div>
                </div>
              }
            </div>
          </div>
          <div className="row mt-4">
            <div className="col">
              <H2>Full Report</H2>
              {!error && !loading &&
                <div className="row">
                  {lastUpdatedDate &&
                    <div className="col-12 col-md-6 py-3">
                      Showing data as of {formatTimestamp({ timestamp: lastUpdatedDate })}
                    </div>
                  }
                  <div className="col-12 col-md-6 text-md-right">
                    <Button
                      label={
                        <span>
                          <span className="fa fa-download" /> Download full report (CSV)
                        </span>
                      }
                      onClick={() => downloadCsv(enterpriseId)}
                      className={['btn-outline-primary']}
                    />
                  </div>
                </div>
              }
              <CourseEnrollmentsTable />
            </div>
          </div>
          <div className="row">
            <div className="col">
              <p>
                For more information, contact edX Enterprise Support at <MailtoLink to="customersuccess@edx.org" content=" customersuccess@edx.org" />.
              </p>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

Admin.defaultProps = {
  error: null,
  loading: false,
  courseCompletions: null,
  activeLearners: null,
  numberOfUsers: null,
  enrolledLearners: null,
  enterpriseId: null,
  lastUpdatedDate: null,
};

Admin.propTypes = {
  getDashboardAnalytics: PropTypes.func.isRequired,
  downloadCsv: PropTypes.func.isRequired,
  enterpriseId: PropTypes.string,
  activeLearners: PropTypes.shape({
    past_week: PropTypes.number,
    past_month: PropTypes.number,
  }),
  enrolledLearners: PropTypes.number,
  numberOfUsers: PropTypes.number,
  courseCompletions: PropTypes.number,
  lastUpdatedDate: PropTypes.string,
  loading: PropTypes.bool,
  error: PropTypes.instanceOf(Error),
};

export default Admin;
