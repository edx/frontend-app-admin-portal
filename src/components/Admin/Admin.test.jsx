import React from 'react';
import PropTypes from 'prop-types';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import Admin from './index';

const mockStore = configureMockStore([thunk]);

class ContextProvider extends React.Component {
  static childContextTypes = {
    store: PropTypes.object.isRequired,
  }

  getChildContext = () => ({
    store: mockStore({
      portalConfiguration: {
        enterpriseId: 'test-enterprise-id',
      },
      courseEnrollments: {
        enrollments: null,
      },
    }),
  })

  render() {
    return this.props.children;
  }
}

ContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

const AdminWrapper = props => (
  <MemoryRouter>
    <ContextProvider>
      <Admin
        getDashboardAnalytics={() => {}}
        downloadCsv={() => {}}
        {...props}
      />
    </ContextProvider>
  </MemoryRouter>
);

describe('<Admin />', () => {
  describe('renders correctly', () => {
    it('calls getDashboardAnalytics prop', () => {
      const mockGetDashboardAnalytics = jest.fn();
      const tree = renderer
        .create((
          <AdminWrapper
            getDashboardAnalytics={mockGetDashboardAnalytics}
            enterpriseId="test-enterprise-id"
          />
        ))
        .toJSON();
      expect(mockGetDashboardAnalytics).toHaveBeenCalled();
      expect(tree).toMatchSnapshot();
    });

    it('with dashboard analytics data', () => {
      const tree = renderer
        .create((
          <AdminWrapper
            activeLearners={{
              past_week: 1,
              past_month: 1,
            }}
            enrolledLearners={1}
            courseCompletions={1}
          />
        ))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('with error state', () => {
      const tree = renderer
        .create((
          <AdminWrapper error={Error('Network Error')} />
        ))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('with loading state', () => {
      const tree = renderer
        .create((
          <AdminWrapper loading />
        ))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('handle changes to enterpriseId prop', () => {
    it('handles non-empty change in enterpriseId prop', () => {
      const mockGetDashboardAnalytics = jest.fn();
      const wrapper = mount((
        <AdminWrapper
          getDashboardAnalytics={mockGetDashboardAnalytics}
          enterpriseId="test-enterprise-id"
        />
      ));

      wrapper.setProps({
        enterpriseId: 'test-enterprise-id-2',
      });

      expect(wrapper.prop('enterpriseId')).toEqual('test-enterprise-id-2');
      expect(mockGetDashboardAnalytics).toBeCalled();
    });

    it('handles empty change in enterpriseId prop', () => {
      const mockGetDashboardAnalytics = jest.fn();
      const wrapper = mount((
        <AdminWrapper
          getDashboardAnalytics={mockGetDashboardAnalytics}
          enterpriseId="test-enterprise-id"
        />
      ));

      wrapper.setProps({
        enterpriseId: null,
      });

      expect(wrapper.prop('enterpriseId')).toEqual(null);
      expect(mockGetDashboardAnalytics).toBeCalled();
    });
  });
});
