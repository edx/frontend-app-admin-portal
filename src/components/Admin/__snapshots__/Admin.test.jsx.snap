// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Admin /> renders correctly calls getDashboardAnalytics prop 1`] = `
<div>
  <div
    className="hero py-4 pt-md-5 pb-md-0"
  >
    <div
      className="container"
    >
      <div
        className="row"
      >
        <div
          className="col-12 col-md-9 mb-2 mb-md-1"
        >
          <h1>
            Learner and Progress Report
          </h1>
        </div>
        <div
          className="col-12 col-md-3 text-md-right mb-2 mb-md-1"
        >
          <img
            alt="edX for Business"
            src="test-file-stub"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="container"
  >
    <div
      className="row mt-4"
    >
      <div
        className="col"
      >
        <h2>
          Overview
        </h2>
      </div>
    </div>
    <div
      className="row mt-4"
    >
      <div
        className="col"
      >
        <h2>
          Full Report
        </h2>
        <div
          className="row"
        >
          <div
            className="col-12 col-md-6 py-3 invisible"
          >
            Showing data as of Month DD, YYYY
          </div>
          <div
            className="col-12 col-md-6 text-md-right"
          >
            <button
              className="btn btn-outline-primary"
              onBlur={[Function]}
              onClick={[Function]}
              onKeyDown={[Function]}
              type="button"
            >
              <span>
                <span
                  className="fa fa-download"
                />
                 Download full report (CSV)
              </span>
            </button>
          </div>
        </div>
        <div />
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col"
      >
        <p>
          For more information, contact edX Enterprise Support at 
          <a
            href="mailto:enterprise@edx.org"
            onClick={null}
            target="_self"
          >
             enterprise@edx.org
          </a>
          .
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Admin /> renders correctly with dashboard analytics data 1`] = `
<div>
  <div
    className="hero py-4 pt-md-5 pb-md-0"
  >
    <div
      className="container"
    >
      <div
        className="row"
      >
        <div
          className="col-12 col-md-9 mb-2 mb-md-1"
        >
          <h1>
            Learner and Progress Report
          </h1>
        </div>
        <div
          className="col-12 col-md-3 text-md-right mb-2 mb-md-1"
        >
          <img
            alt="edX for Business"
            src="test-file-stub"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="container"
  >
    <div
      className="row mt-4"
    >
      <div
        className="col"
      >
        <h2>
          Overview
        </h2>
        <div
          className="row mt-3 equal-col-height"
        >
          <div
            className="col-xs-12 col-md-6 col-xl-3 mb-3"
          >
            <div
              className="card number-card"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title d-flex align-items-center justify-content-between"
                >
                  <span>
                    X
                  </span>
                  <div>
                    <span
                      aria-hidden={true}
                      className="fa fa-users d-flex align-items-center justify-content-center"
                      id="Icon1"
                    />
                  </div>
                </h5>
                <p
                  className="card-text"
                >
                  total number of learners registered
                </p>
              </div>
            </div>
          </div>
          <div
            className="col-xs-12 col-md-6 col-xl-3 mb-3"
          >
            <div
              className="card number-card"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title d-flex align-items-center justify-content-between"
                >
                  <span>
                    1
                  </span>
                  <div>
                    <span
                      aria-hidden={true}
                      className="fa fa-check d-flex align-items-center justify-content-center"
                      id="Icon1"
                    />
                  </div>
                </h5>
                <p
                  className="card-text"
                >
                  learners enrolled in at least one course
                </p>
              </div>
            </div>
          </div>
          <div
            className="col-xs-12 col-md-6 col-xl-3 mb-3"
          >
            <div
              className="card number-card"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title d-flex align-items-center justify-content-between"
                >
                  <span>
                    1
                  </span>
                  <div>
                    <span
                      aria-hidden={true}
                      className="fa fa-eye d-flex align-items-center justify-content-center"
                      id="Icon1"
                    />
                  </div>
                </h5>
                <p
                  className="card-text"
                >
                  active learners in the past week
                </p>
              </div>
            </div>
          </div>
          <div
            className="col-xs-12 col-md-6 col-xl-3 mb-3"
          >
            <div
              className="card number-card"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title d-flex align-items-center justify-content-between"
                >
                  <span>
                    1
                  </span>
                  <div>
                    <span
                      aria-hidden={true}
                      className="fa fa-trophy d-flex align-items-center justify-content-center"
                      id="Icon1"
                    />
                  </div>
                </h5>
                <p
                  className="card-text"
                >
                  course completions
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="row mt-4"
    >
      <div
        className="col"
      >
        <h2>
          Full Report
        </h2>
        <div
          className="row"
        >
          <div
            className="col-12 col-md-6 py-3 invisible"
          >
            Showing data as of Month DD, YYYY
          </div>
          <div
            className="col-12 col-md-6 text-md-right"
          >
            <button
              className="btn btn-outline-primary"
              onBlur={[Function]}
              onClick={[Function]}
              onKeyDown={[Function]}
              type="button"
            >
              <span>
                <span
                  className="fa fa-download"
                />
                 Download full report (CSV)
              </span>
            </button>
          </div>
        </div>
        <div />
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col"
      >
        <p>
          For more information, contact edX Enterprise Support at 
          <a
            href="mailto:enterprise@edx.org"
            onClick={null}
            target="_self"
          >
             enterprise@edx.org
          </a>
          .
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Admin /> renders correctly with error state 1`] = `
<div>
  <div
    className="hero py-4 pt-md-5 pb-md-0"
  >
    <div
      className="container"
    >
      <div
        className="row"
      >
        <div
          className="col-12 col-md-9 mb-2 mb-md-1"
        >
          <h1>
            Learner and Progress Report
          </h1>
        </div>
        <div
          className="col-12 col-md-3 text-md-right mb-2 mb-md-1"
        >
          <img
            alt="edX for Business"
            src="test-file-stub"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="container"
  >
    <div
      className="row mt-4"
    >
      <div
        className="col"
      >
        <h2>
          Overview
        </h2>
        <div
          className="mt-3 alert fade alert-danger show"
          hidden={false}
          role="alert"
        >
          <div
            className="alert-dialog"
          >
            <div
              className="d-flex"
            >
              <div
                className="icon mr-2"
              >
                <div>
                  <span
                    aria-hidden={true}
                    className="fa fa-times-circle"
                    id="Icon1"
                  />
                </div>
              </div>
              <div
                className="message"
              >
                <span
                  className="title"
                >
                  Unable to load overview
                </span>
                <span>
                  Try refreshing your screen (Network Error)
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="row mt-4"
    >
      <div
        className="col"
      >
        <h2>
          Full Report
        </h2>
        <div
          className="row"
        >
          <div
            className="col-12 col-md-6 py-3 invisible"
          >
            Showing data as of Month DD, YYYY
          </div>
          <div
            className="col-12 col-md-6 text-md-right"
          >
            <button
              className="btn btn-outline-primary"
              onBlur={[Function]}
              onClick={[Function]}
              onKeyDown={[Function]}
              type="button"
            >
              <span>
                <span
                  className="fa fa-download"
                />
                 Download full report (CSV)
              </span>
            </button>
          </div>
        </div>
        <div />
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col"
      >
        <p>
          For more information, contact edX Enterprise Support at 
          <a
            href="mailto:enterprise@edx.org"
            onClick={null}
            target="_self"
          >
             enterprise@edx.org
          </a>
          .
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Admin /> renders correctly with loading state 1`] = `
<div>
  <div
    className="hero py-4 pt-md-5 pb-md-0"
  >
    <div
      className="container"
    >
      <div
        className="row"
      >
        <div
          className="col-12 col-md-9 mb-2 mb-md-1"
        >
          <h1>
            Learner and Progress Report
          </h1>
        </div>
        <div
          className="col-12 col-md-3 text-md-right mb-2 mb-md-1"
        >
          <img
            alt="edX for Business"
            src="test-file-stub"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="container"
  >
    <div
      className="row mt-4"
    >
      <div
        className="col"
      >
        <h2>
          Overview
        </h2>
        <div
          className="overview mt-3 loading d-flex align-items-center justify-content-center"
        >
          <span>
            Loading...
          </span>
        </div>
      </div>
    </div>
    <div
      className="row mt-4"
    >
      <div
        className="col"
      >
        <h2>
          Full Report
        </h2>
        <div
          className="row"
        >
          <div
            className="col-12 col-md-6 py-3 invisible"
          >
            Showing data as of Month DD, YYYY
          </div>
          <div
            className="col-12 col-md-6 text-md-right"
          >
            <button
              className="btn btn-outline-primary"
              onBlur={[Function]}
              onClick={[Function]}
              onKeyDown={[Function]}
              type="button"
            >
              <span>
                <span
                  className="fa fa-download"
                />
                 Download full report (CSV)
              </span>
            </button>
          </div>
        </div>
        <div />
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col"
      >
        <p>
          For more information, contact edX Enterprise Support at 
          <a
            href="mailto:enterprise@edx.org"
            onClick={null}
            target="_self"
          >
             enterprise@edx.org
          </a>
          .
        </p>
      </div>
    </div>
  </div>
</div>
`;
